File: ptr_c/opaque.c
//zig cc -c opaque.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Op {
    /*char name[32];*/
    const char *name;
    const char **names;
    int size;
} op_t;

void* op_new(const char *name) {
    op_t *pp = malloc(sizeof(op_t));
    pp->name = name;
    /*memset(pp, 0, sizeof(op_t));*/
    /*if(!pp) {*/
        /*strcpy(pp->name, name);*/
    /*}*/
    return (void*)pp;
}
op_t* op_new_all(const char **names, int size) {
    op_t *pp = malloc(sizeof(op_t));
    pp->names = names;
    pp->size = size;
    return pp;
}
void op_free(op_t *pp) {
    free(pp);
}
void hello(op_t *pp) {
    printf("%s\n", pp->name);
}
void hello_all(op_t *pp) {
    for(int i=0; i<pp->size; ++i) {
        printf("%s\n", pp->names[i]);
    }
}

File: ptr_c/opaque.zig
//zig cc -c opaque.c
//zig run opaque.zig opaque.o
const std = @import("std");

//pub const struct_Op = opaque {};
pub const struct_Op = extern struct {
    name: [*c]u8,
    names: [*c][*c]const u8,
    size: u32,
};

pub const op_t = struct_Op;
pub const op_p = ?*op_t;
pub extern fn op_new(name: [*]const u8) op_p;
//pub extern fn op_new(name: [*]const u8) ?*anyopaque;
pub extern fn op_new_all(names: [*c][*]const u8, size: u32) op_p;
//pub extern fn op_free(pp: op_p) void;
pub extern fn op_free(pp: op_p) void;
//pub extern fn hello(pp: op_p) void;
//pub extern fn hello_all(pp: op_p) void;
pub extern fn hello(pp: ?*anyopaque) void;
pub extern fn hello_all(pp: ?*anyopaque) void;

pub fn main() !void {
    {
        const pp = op_new("world");
        hello(pp);
        op_free(pp);
    }
    {
        var names = [_][*]const u8{"hello", "world"};
        _ = &names;
        const pp = op_new_all(names[0..], 2);
        //const pp = op_new_all(names[0..].ptr, 2);

        //var names = [_][]const u8{"hello", "world"};
        //var names_c: [2][*]const u8 = undefined;
        //names_c[0] = names[0].ptr;
        //names_c[1] = names[1].ptr;
        //const pp = op_new_all(names[0..].ptr, 2);
        //const pp = op_new_all(names_c[0..], 2);
        pp.?.names[0] = "test";
        hello_all(pp);
        op_free(pp);
    }
}

